# **ETL de Datasets provistos**

### Importamos las librerías necesarias

In [203]:
! pip install dbfread

Defaulting to user installation because normal site-packages is not writeable


In [204]:
import pandas as pd
import dbfread
from re import search
import warnings
warnings.filterwarnings("ignore")

## ETL: `Electric and Alternative Fuel Charging Stations.csv`

In [205]:
# Leemos el dataset
electric_fuel_charg_stations = pd.read_csv(r'C:\Users\melan\Desktop\PG 1\ETL -EDA\Data\1. Originales\Electric and Alternative Fuel Charging Stations.csv')
electric_fuel_charg_stations.head(2)

Fuel Type Code                          Station Name    Street Address  \
0            CNG  Spire - Montgomery Operations Center  2951 Chestnut St   
1            CNG                   PS Energy - Atlanta  340 Whitehall St   

                             Intersection Directions        City State    ZIP  \
0                                                NaN  Montgomery    AL  36107   
1  From I-7585 N, exit 91 to Central Ave, left on...     Atlanta    GA  30303   

   Plus4 Station Phone Status Code  ... EV Pricing (French) LPG Nozzle Types  \
0    NaN           NaN           E  ...                 NaN              NaN   
1    NaN  770-350-3000           E  ...                 NaN              NaN   

  Hydrogen Pressures Hydrogen Standards CNG Fill Type Code CNG PSI  \
0                NaN                NaN                  B    3600   
1                NaN                NaN                  Q    3600   

  CNG Vehicle Class  LNG Vehicle Class  EV On-Site Renewable Source  \
0                MD                NaN                          NaN   
1                MD                NaN                          NaN   

   Restricted Access  
0                NaN  
1              False  

[2 rows x 65 columns]

In [206]:
# Revisamos las columnas para evaluar su contenido

electric_fuel_charg_stations.columns

Index(['Fuel Type Code', 'Station Name', 'Street Address',
       'Intersection Directions', 'City', 'State', 'ZIP', 'Plus4',
       'Station Phone', 'Status Code', 'Expected Date',
       'Groups With Access Code', 'Access Days Time', 'Cards Accepted',
       'BD Blends', 'NG Fill Type Code', 'NG PSI', 'EV Level1 EVSE Num',
       'EV Level2 EVSE Num', 'EV DC Fast Count', 'EV Other Info', 'EV Network',
       'EV Network Web', 'Geocode Status', 'Latitude', 'Longitude',
       'Date Last Confirmed', 'ID', 'Updated At', 'Owner Type Code',
       'Federal Agency ID', 'Federal Agency Name', 'Open Date',
       'Hydrogen Status Link', 'NG Vehicle Class', 'LPG Primary',
       'E85 Blender Pump', 'EV Connector Types', 'Country',
       'Intersection Directions (French)', 'Access Days Time (French)',
       'BD Blends (French)', 'Groups With Access Code (French)',
       'Hydrogen Is Retail', 'Access Code', 'Access Detail Code',
       'Federal Agency Code', 'Facility Type', 'CNG Dispenser N

In [207]:
# Eliminamos las columnas que no importantes
eliminar = ['Intersection Directions', 'ZIP', 'Plus4','Restricted Access',
       'Station Phone', 'Status Code', 'Expected Date',
       'Groups With Access Code', 'Access Days Time', 'Cards Accepted',
       'BD Blends', 'NG Fill Type Code', 'NG PSI', 'EV Level1 EVSE Num',
       'EV Level2 EVSE Num', 'EV DC Fast Count', 'EV Other Info', 'EV Network',
       'EV Network Web', 'Geocode Status','EV On-Site Renewable Source',
       'Date Last Confirmed', 'ID', 'Updated At', 'Owner Type Code',
       'Federal Agency ID', 'Federal Agency Name', 'Open Date',
       'Hydrogen Status Link', 'NG Vehicle Class', 'LPG Primary',
       'E85 Blender Pump', 'EV Connector Types', 'Country','LNG Vehicle Class',
       'Intersection Directions (French)', 'Access Days Time (French)',
       'BD Blends (French)', 'Groups With Access Code (French)',
       'Hydrogen Is Retail', 'Access Code', 'Access Detail Code',
       'Federal Agency Code', 'Facility Type', 'CNG Dispenser Num',
       'CNG On-Site Renewable Source', 'CNG Total Compression Capacity',
       'CNG Storage Capacity', 'LNG On-Site Renewable Source',
       'E85 Other Ethanol Blends', 'EV Pricing', 'EV Pricing (French)',
       'LPG Nozzle Types', 'Hydrogen Pressures', 'Hydrogen Standards',
       'CNG Fill Type Code', 'CNG PSI', 'CNG Vehicle Class']

electric_fuel_charg_stations.drop(eliminar, axis=1, inplace=True)
electric_fuel_charg_stations.head(2)

Fuel Type Code                          Station Name    Street Address  \
0            CNG  Spire - Montgomery Operations Center  2951 Chestnut St   
1            CNG                   PS Energy - Atlanta  340 Whitehall St   

         City State   Latitude  Longitude  
0  Montgomery    AL  32.367916 -86.267021  
1     Atlanta    GA  33.745843 -84.398837

In [208]:
electric_fuel_charg_stations = electric_fuel_charg_stations[electric_fuel_charg_stations['State'] == 'NY']
electric_fuel_charg_stations

Fuel Type Code                               Station Name  \
7                CNG  Clean Energy - Greenpoint - National Grid   
8                CNG                   Canarsie - National Grid   
9                CNG      Con Edison - W 29th St Service Center   
10               CNG      Con Edison - E 16th St Service Center   
11               CNG       Con Edison - Van Nest Service Center   
...              ...                                        ...   
70266           ELEC                       GENESEO W PARK RIGHT   
70267           ELEC                      GENESEO W PARK CENTER   
70278           ELEC                   G & M REALTY 2 STATION 1   
70288           ELEC                     40 N 4th St Apartments   
70395           ELEC                         EVolveNY Salamanca   

                     Street Address       City State   Latitude  Longitude  
7                   287 Maspeth Ave   Brooklyn    NY  40.718037 -73.932309  
8                   8424 Ditmas Ave   Brooklyn    NY  40.645540 -73.918344  
9                      281 11th Ave   New York    NY  40.752903 -74.005831  
10                    700 E 16th St   New York    NY  40.729175 -73.973310  
11               1615 Bronxdale Ave      Bronx    NY  40.844188 -73.861572  
...                             ...        ...   ...        ...        ...  
70266  4271 Genesee Valley Plaza Rd    Geneseo    NY  42.796997 -77.785406  
70267  4271 Genesee Valley Plaza Rd    Geneseo    NY  42.796960 -77.785320  
70278             22-44 Jackson Ave        LIC    NY  40.745393 -73.946234  
70288                   40 N 4th ST   Brooklyn    NY  40.718391 -73.963339  
70395              768 Broad Street  Salamanca    NY  42.160008 -78.744875  

[3409 rows x 7 columns]

In [209]:
# Mantenemos solo las filas donde City = Manhattan, Brooklyn, Bronx, Queens, Staten Island
electric_fuel_charg_stations = electric_fuel_charg_stations[electric_fuel_charg_stations['City'].isin(['Manhattan', 'Brooklyn', 'Bronx', 'Queens', 'Staten Island'])]
electric_fuel_charg_stations

Fuel Type Code                               Station Name  \
7                CNG  Clean Energy - Greenpoint - National Grid   
8                CNG                   Canarsie - National Grid   
11               CNG       Con Edison - Van Nest Service Center   
13               CNG  Con Edison - College Point Service Center   
1500             CNG                                  Bronx Zoo   
...              ...                                        ...   
69624           ELEC                   Rapid Park - 44 State St   
69640           ELEC                               906 Union St   
69993           ELEC                            Supercharge NYC   
70050           ELEC                            35 Red Cross Pl   
70288           ELEC                     40 N 4th St Apartments   

           Street Address      City State   Latitude  Longitude  
7         287 Maspeth Ave  Brooklyn    NY  40.718037 -73.932309  
8         8424 Ditmas Ave  Brooklyn    NY  40.645540 -73.918344  
11     1615 Bronxdale Ave     Bronx    NY  40.844188 -73.861572  
13        124-15 31st Ave    Queens    NY  40.769623 -73.843069  
1500   2300 Southern Blvd     Bronx    NY  40.850018 -73.878128  
...                   ...       ...   ...        ...        ...  
69624         44 State St  Brooklyn    NY  40.691800 -73.998290  
69640        906 Union St  Brooklyn    NY  40.672830 -73.971750  
69993       1074 Grand St  Brooklyn    NY  40.713604 -73.933099  
70050     35 Red Cross Pl  Brooklyn    NY  40.699487 -73.989516  
70288         40 N 4th ST  Brooklyn    NY  40.718391 -73.963339  

[200 rows x 7 columns]

In [210]:
# Renombramos las columnas para mejor identificacion
electric_fuel_charg_stations = electric_fuel_charg_stations.copy()
electric_fuel_charg_stations.rename(columns={'Fuel Type Code':'Fuel Code', 'City':'Borough'}, inplace=True)
electric_fuel_charg_stations.head(2)

Fuel Code                               Station Name   Street Address  \
7       CNG  Clean Energy - Greenpoint - National Grid  287 Maspeth Ave   
8       CNG                   Canarsie - National Grid  8424 Ditmas Ave   

    Borough State   Latitude  Longitude  
7  Brooklyn    NY  40.718037 -73.932309  
8  Brooklyn    NY  40.645540 -73.918344

In [211]:
# Eliminamos la columna State porque no aporta nada
electric_fuel_charg_stations.drop('State', axis=1, inplace=True)
electric_fuel_charg_stations.head(2)

Fuel Code                               Station Name   Street Address  \
7       CNG  Clean Energy - Greenpoint - National Grid  287 Maspeth Ave   
8       CNG                   Canarsie - National Grid  8424 Ditmas Ave   

    Borough   Latitude  Longitude  
7  Brooklyn  40.718037 -73.932309  
8  Brooklyn  40.645540 -73.918344

In [212]:
electric_fuel_charg_stations['Fuel Code'].unique()

array(['CNG', 'ELEC', 'E85', 'BD'], dtype=object)

### Extraemos y cargo el archivo en formato parquet

In [213]:
electric_fuel_charg_stations.to_parquet(r'C:\Users\melan\Desktop\PG 1\ETL -EDA\Data\2. Normalizacion\electric_and_alternative_fuel_charging_stations.parquet', index=False)


## ETL: `ElectricCarData_Norm.csv`

In [214]:
# Leemos el dataset para validar su contenido y tratar sus datos
electric_cardata = pd.read_csv(r'C:\Users\melan\Desktop\PG 1\ETL -EDA\Data\1. Originales\ElectricCarData_Norm.csv')

electric_cardata.head(2)

Brand                          Model     Accel  TopSpeed   Range  \
0       Tesla   Model 3 Long Range Dual Motor   4.6 sec  233 km/h  450 km   
1  Volkswagen                       ID.3 Pure  10.0 sec  160 km/h  270 km   

  Efficiency FastCharge              RapidCharge        PowerTrain  \
0  161 Wh/km   940 km/h  Rapid charging possible   All Wheel Drive   
1  167 Wh/km   250 km/h  Rapid charging possible  Rear Wheel Drive   

     PlugType  BodyStyle Segment  Seats  PriceEuro  
0  Type 2 CCS      Sedan       D      5      55480  
1  Type 2 CCS  Hatchback       C      5      30000

In [215]:
# Revisamos las columnas 
electric_cardata.columns

Index(['Brand', 'Model', 'Accel', 'TopSpeed', 'Range', 'Efficiency',
       'FastCharge', 'RapidCharge', 'PowerTrain', 'PlugType', 'BodyStyle',
       'Segment', 'Seats', 'PriceEuro'],
      dtype='object')

In [216]:
# Eliminamos las columnas que no suman
eliminar = ['Accel','TopSpeed','PowerTrain','PlugType','Segment','Seats']
electric_cardata.drop(eliminar, axis=1, inplace=True)
electric_cardata.head(2)

Brand                          Model   Range Efficiency FastCharge  \
0       Tesla   Model 3 Long Range Dual Motor  450 km  161 Wh/km   940 km/h   
1  Volkswagen                       ID.3 Pure  270 km  167 Wh/km   250 km/h   

               RapidCharge  BodyStyle  PriceEuro  
0  Rapid charging possible      Sedan      55480  
1  Rapid charging possible  Hatchback      30000

In [217]:
# Renombramos las columnas 
electric_cardata = electric_cardata.copy()
electric_cardata.rename(columns={'Brand':'Manufacturer','BodyStyle':'Category','PriceEuro':'Price'}, inplace=True)
electric_cardata.head(2)

Manufacturer                          Model   Range Efficiency FastCharge  \
0       Tesla   Model 3 Long Range Dual Motor  450 km  161 Wh/km   940 km/h   
1  Volkswagen                       ID.3 Pure  270 km  167 Wh/km   250 km/h   

               RapidCharge   Category  Price  
0  Rapid charging possible      Sedan  55480  
1  Rapid charging possible  Hatchback  30000

In [218]:
# Revisamos el contenido de la columna Category para validar informacion relevante
electric_cardata.Category.unique()

array(['Sedan', 'Hatchback', 'Liftback', 'SUV', 'Pickup', 'MPV', 'Cabrio',
       'SPV', 'Station'], dtype=object)

In [219]:
# Mantenemos solo las filas donde Category = Sedan, SUV
electric_cardata = electric_cardata[electric_cardata['Category'].isin(['Sedan', 'SUV'])]
electric_cardata.head(2)

Manufacturer                          Model   Range Efficiency FastCharge  \
0       Tesla   Model 3 Long Range Dual Motor  450 km  161 Wh/km   940 km/h   
3         BMW                            iX3   360 km  206 Wh/km   560 km/h   

               RapidCharge Category  Price  
0  Rapid charging possible    Sedan  55480  
3  Rapid charging possible      SUV  68040

In [220]:
electric_cardata.dtypes

Manufacturer    object
Model           object
Range           object
Efficiency      object
FastCharge      object
RapidCharge     object
Category        object
Price            int64
dtype: object

In [221]:
# Cambiamos el nombre de las columnas 'Range', 'Efficiency' y 'FastCharge' para agregarles la unidad.
electric_cardata = electric_cardata.rename(columns={'Range':'Range (mile)', 'Efficiency':'Efficiency (Wh/mile)', 'FastCharge':'FastCharge (mile/h)'})
electric_cardata.head(2)

Manufacturer                          Model Range (mile)  \
0       Tesla   Model 3 Long Range Dual Motor       450 km   
3         BMW                            iX3        360 km   

  Efficiency (Wh/mile) FastCharge (mile/h)              RapidCharge Category  \
0            161 Wh/km            940 km/h  Rapid charging possible    Sedan   
3            206 Wh/km            560 km/h  Rapid charging possible      SUV   

   Price  
0  55480  
3  68040

In [222]:
# Define the function to extract numeric values
def extraer_valor_numerico(value):
    try:
        # Split the value by whitespace and extract the first part
        numeric_value = value.split()[0]
        # Convert the extracted part to float
        numeric_value = float(numeric_value)
        return numeric_value
    except:
        # If any error occurs during conversion, return None
        return None

# Apply the function to extract numeric values from specific columns
electric_cardata['Range (mile)'] = electric_cardata['Range (mile)'].apply(extraer_valor_numerico)
electric_cardata['Efficiency (Wh/mile)'] = electric_cardata['Efficiency (Wh/mile)'].apply(extraer_valor_numerico)
electric_cardata['FastCharge (mile/h)'] = electric_cardata['FastCharge (mile/h)'].apply(extraer_valor_numerico)


In [223]:
# Define the function to convert kilometers to miles
def convert_km_to_mile(km):
    # Conversion factor from kilometers to miles
    conversion_factor = 0.621371
    # Convert kilometers to miles
    mile = km * conversion_factor
    return mile

# Aplicamos la funcion de `convert_km_to_mile` a las columnas 'Range', 'Efficiency' y 'FastCharge'
electric_cardata['Range (mile)'] = electric_cardata['Range (mile)'].apply(convert_km_to_mile)
electric_cardata['Efficiency (Wh/mile)'] = electric_cardata['Efficiency (Wh/mile)'].apply(convert_km_to_mile)
electric_cardata['FastCharge (mile/h)'] = electric_cardata['FastCharge (mile/h)'].apply(convert_km_to_mile)


In [224]:
electric_cardata.head(2)

Manufacturer                          Model  Range (mile)  \
0       Tesla   Model 3 Long Range Dual Motor     279.61695   
3         BMW                            iX3      223.69356   

   Efficiency (Wh/mile)  FastCharge (mile/h)              RapidCharge  \
0            100.040731            584.08874  Rapid charging possible   
3            128.002426            347.96776  Rapid charging possible   

  Category  Price  
0    Sedan  55480  
3      SUV  68040

## ETL: `Light Duty Vehicles.csv`

In [225]:
# Leemos el dataset y validamos datos
light_duty_veh = pd.read_csv(r'C:\Users\melan\Desktop\PG 1\ETL -EDA\Data\1. Originales\Light Duty Vehicles.csv')
light_duty_veh.head(2)

Vehicle ID  Fuel ID  Fuel Configuration ID  Manufacturer ID  Category ID  \
0       13044       45                    9.0              365           27   
1       12854       45                    9.0              377           27   

  Model  Model Year  Alternative Fuel Economy City  \
0   NSX        2022                            NaN   
1   A3         2022                            NaN   

   Alternative Fuel Economy Highway  Alternative Fuel Economy Combined  ...  \
0                               NaN                                NaN  ...   
1                               NaN                                NaN  ...   

               Manufacturer URL     Category  Fuel Code             Fuel  \
0         http://www.acura.com/  Sedan/Wagon       HYBR  Hybrid Electric   
1  http://progress.audiusa.com/  Sedan/Wagon       HYBR  Hybrid Electric   

  Fuel Configuration Name Electric-Only Range  PHEV Total Range PHEV Type  \
0         Hybrid Electric                 NaN               NaN       NaN   
1         Hybrid Electric                 NaN               NaN       NaN   

  Notes Drivetrain  
0   NaN        AWD  
1   NaN        FWD  

[2 rows x 29 columns]

In [226]:
# Revisamos cantidad de nulos y nombre de las columnas
light_duty_veh.isnull().sum()

Vehicle ID                               0
Fuel ID                                  0
Fuel Configuration ID                   35
Manufacturer ID                          0
Category ID                              0
Model                                    0
Model Year                               0
Alternative Fuel Economy City         1413
Alternative Fuel Economy Highway      1503
Alternative Fuel Economy Combined     2541
Conventional Fuel Economy City         964
Conventional Fuel Economy Highway      962
Conventional Fuel Economy Combined    2245
Transmission Type                       86
Engine Type                            797
Engine Size                            134
Engine Cylinder Count                  565
Engine Description                    1037
Manufacturer                             0
Manufacturer URL                       164
Category                                 0
Fuel Code                                0
Fuel                                     0
Fuel Config

In [227]:
light_duty_veh.columns

Index(['Vehicle ID', 'Fuel ID', 'Fuel Configuration ID', 'Manufacturer ID',
       'Category ID', 'Model', 'Model Year', 'Alternative Fuel Economy City',
       'Alternative Fuel Economy Highway', 'Alternative Fuel Economy Combined',
       'Conventional Fuel Economy City', 'Conventional Fuel Economy Highway',
       'Conventional Fuel Economy Combined', 'Transmission Type',
       'Engine Type', 'Engine Size', 'Engine Cylinder Count',
       'Engine Description', 'Manufacturer', 'Manufacturer URL', 'Category',
       'Fuel Code', 'Fuel', 'Fuel Configuration Name', 'Electric-Only Range',
       'PHEV Total Range', 'PHEV Type', 'Notes', 'Drivetrain'],
      dtype='object')

In [228]:
# Eliminamos las columnas que no nos interesan
eliminar = ['Vehicle ID', 'Fuel ID', 'Fuel Configuration ID', 'Manufacturer ID',
            'Category ID', 'Alternative Fuel Economy City',
            'Alternative Fuel Economy Highway', 'Alternative Fuel Economy Combined',
            'Conventional Fuel Economy City', 'Conventional Fuel Economy Highway',
            'Conventional Fuel Economy Combined', 'Transmission Type',
            'Engine Type', 'Engine Size', 'Engine Cylinder Count',
            'Engine Description', 'Manufacturer URL',
            'Fuel Code', 'Fuel Configuration Name', 'Electric-Only Range',
            'PHEV Total Range', 'PHEV Type', 'Notes', 'Drivetrain']

light_duty_veh.drop(eliminar, axis=1, inplace=True)
light_duty_veh.head(2)

Model  Model Year Manufacturer     Category             Fuel
0   NSX        2022        Acura  Sedan/Wagon  Hybrid Electric
1   A3         2022         Audi  Sedan/Wagon  Hybrid Electric

In [229]:
# Renombramos las columnas
light_duty_veh = light_duty_veh.copy()
light_duty_veh.rename(columns={'Model Year':'Year'}, inplace=True)
light_duty_veh.head(2)

Model  Year Manufacturer     Category             Fuel
0   NSX  2022        Acura  Sedan/Wagon  Hybrid Electric
1   A3   2022         Audi  Sedan/Wagon  Hybrid Electric

In [230]:
# Revisamos la composicion de la columna Category
light_duty_veh.Category.unique()

array(['Sedan/Wagon', 'SUV', 'Pickup', 'Van', 'Passenger Van/Shuttle Bus',
       'Vocational/Cab Chassis'], dtype=object)

In [231]:
# Mantenemos solo las filas donde Category = Sedan/Wagon, SUV
light_duty_veh = light_duty_veh[light_duty_veh['Category'].isin(['Sedan/Wagon', 'SUV'])]
light_duty_veh.head(2)

Model  Year Manufacturer     Category             Fuel
0   NSX  2022        Acura  Sedan/Wagon  Hybrid Electric
1   A3   2022         Audi  Sedan/Wagon  Hybrid Electric

## ETL: `taxi_zones.dbf`

In [232]:
# Leemos el dataset 
taxi_zones = dbfread.DBF(r'C:\Users\melan\Desktop\PG 1\ETL -EDA\Data\1. Originales\taxi_zones.dbf')
taxi_zones = pd.DataFrame(iter(taxi_zones))
taxi_zones

OBJECTID  Shape_Leng  Shape_Area                     zone  LocationID  \
0           1    0.116357    0.000782           Newark Airport           1   
1           2    0.433470    0.004866              Jamaica Bay           2   
2           3    0.084341    0.000314  Allerton/Pelham Gardens           3   
3           4    0.043567    0.000112            Alphabet City           4   
4           5    0.092146    0.000498            Arden Heights           5   
..        ...         ...         ...                      ...         ...   
258       259    0.126750    0.000395       Woodlawn/Wakefield         259   
259       260    0.133514    0.000422                 Woodside         260   
260       261    0.027120    0.000034       World Trade Center         261   
261       262    0.049064    0.000122           Yorkville East         262   
262       263    0.037017    0.000066           Yorkville West         263   

           borough  
0              EWR  
1           Queens  
2            Bronx  
3        Manhattan  
4    Staten Island  
..             ...  
258          Bronx  
259         Queens  
260      Manhattan  
261      Manhattan  
262      Manhattan  

[263 rows x 6 columns]

In [233]:
# Eliminamos las columnas que no nos interesan
eliminar = ['OBJECTID', 'Shape_Leng', 'Shape_Area']
taxi_zones.drop(eliminar, axis=1, inplace=True)
taxi_zones.head(2)

zone  LocationID borough
0  Newark Airport           1     EWR
1     Jamaica Bay           2  Queens

In [234]:
# Renombramos las columnas
taxi_zones = taxi_zones.copy()
taxi_zones.rename(columns={'zone':'Zone', 'borough':'Borough'}, inplace=True)

taxi_zones.head(2)

Zone  LocationID Borough
0  Newark Airport           1     EWR
1     Jamaica Bay           2  Queens

In [235]:
# Se Define the function to map borough names to IDs
def borough_id(borough_name):
    # Dictionary mapping borough names to IDs
    borough_mapping = {
        'Manhattan': 1,
        'Bronx': 2,
        'Brooklyn': 3,
        'Queens': 4,
        'Staten Island': 5
    }
    # Return the ID corresponding to the borough name
    return borough_mapping.get(borough_name, None)


In [236]:
# Se implementa la funcion 'borough_id' a la columna 'Borough' para crear la nueva columna 'BoroughID'
taxi_zones['BoroughID'] = taxi_zones['Borough'].apply(borough_id)
taxi_zones.head(2)


Zone  LocationID Borough  BoroughID
0  Newark Airport           1     EWR        NaN
1     Jamaica Bay           2  Queens        4.0

In [237]:
# Siguiendo con la lógica implementada, asignamos el valor 6 al BoroughID de valor 0 que corresponde a EWR
taxi_zones.loc[taxi_zones['BoroughID'] == 0, 'BoroughID'] = 6
taxi_zones.head(2)

Zone  LocationID Borough  BoroughID
0  Newark Airport           1     EWR        NaN
1     Jamaica Bay           2  Queens        4.0

### Guardamos el archivo en formato parquet

In [238]:
taxi_zones.to_parquet('C:\\Users\\melan\\Desktop\\PG 1\\ETL -EDA\\Data\\2. Normalizacion\\taxi_zones.parquet', index=False)


## ETL: `Vehicle Fuel Economy Data.csv`

In [239]:
# Leemos el dataset y validamos la info 
vehicle_fuel = pd.read_csv(r'C:\Users\melan\Desktop\PG 1\ETL -EDA\Data\1. Originales\Vehicle Fuel Economy Data.csv')
vehicle_fuel.head(2)

Year Manufacturer         Model  barrels08  barrelsA08  charge240  city08  \
0  2017          BMW      M6 Coupe  17.500588         0.0        0.0    15.0   
1  2017          BMW  X3 sDrive28i  12.396250         0.0        0.0    21.0   

   city08U  cityA08  cityA08U  ...  mfrCode  c240Dscr  charge240b  c240bDscr  \
0     14.0      0.0       0.0  ...      BMX       NaN         0.0        NaN   
1     21.0      0.0       0.0  ...      BMX       NaN         0.0        NaN   

    createdOn  modifiedOn  startStop  phevCity  phevHwy  phevComb  
0  04-04-2016  09-09-2016          Y       0.0      0.0       0.0  
1  04-04-2016  22-11-2016          Y       0.0      0.0       0.0  

[2 rows x 82 columns]

In [240]:
# Revisamos su informacion
vehicle_fuel.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46186 entries, 0 to 46185
Data columns (total 82 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Year             46186 non-null  int64  
 1   Manufacturer     46186 non-null  object 
 2   Model            46176 non-null  object 
 3   barrels08        46169 non-null  float64
 4   barrelsA08       46169 non-null  float64
 5   charge240        46169 non-null  float64
 6   city08           46169 non-null  float64
 7   city08U          46169 non-null  float64
 8   cityA08          46169 non-null  float64
 9   cityA08U         46169 non-null  float64
 10  cityCD           46169 non-null  float64
 11  cityE            46169 non-null  float64
 12  cityUF           46169 non-null  float64
 13  co2              46169 non-null  float64
 14  co2A             46169 non-null  float64
 15  co2TailpipeAGpm  46169 non-null  float64
 16  co2TailpipeGpm   46169 non-null  float64
 17  comb08      

In [241]:
# Revisamos las columnas
vehicle_fuel.columns

Index(['Year', 'Manufacturer', 'Model', 'barrels08', 'barrelsA08', 'charge240',
       'city08', 'city08U', 'cityA08', 'cityA08U', 'cityCD', 'cityE', 'cityUF',
       'co2', 'co2A', 'co2TailpipeAGpm', 'co2TailpipeGpm', 'comb08', 'comb08U',
       'combA08', 'combA08U', 'combE', 'combinedCD', 'combinedUF', 'cylinders',
       'displ', 'drive', 'engId', 'eng_dscr', 'feScore', 'fuelCost08',
       'fuelCostA08', 'fuelType', 'fuelType1', 'ghgScore', 'ghgScoreA',
       'highway08', 'highway08U', 'highwayA08', 'highwayA08U', 'VClass',
       'highwayCD', 'highwayE', 'highwayUF', 'hlv', 'hpv', 'id', 'lv2', 'lv4',
       'mpgData', 'phevBlended', 'pv2', 'pv4', 'range', 'rangeCity',
       'rangeCityA', 'rangeHwy', 'rangeHwyA', 'trany', 'UCity', 'UCityA',
       'UHighway', 'UHighwayA', 'youSaveSpend', 'guzzler', 'trans_dscr',
       'tCharger', 'sCharger', 'atvType', 'fuelType2', 'rangeA', 'evMotor',
       'mfrCode', 'c240Dscr', 'charge240b', 'c240bDscr', 'createdOn',
       'modifiedOn', 's

In [242]:
vehicle_fuel.fuelType.unique()

array(['Premium', 'Regular', 'Diesel', 'Electricity', 'Gasoline or E85',
       'Midgrade', 'Premium and Electricity', 'Premium or E85', nan,
       'Regular Gas and Electricity', 'Regular Gas or Electricity',
       'Premium Gas or Electricity', 'CNG', 'Gasoline or natural gas',
       'Gasoline or propane'], dtype=object)

In [243]:
vehicle_fuel.fuelType1.unique()

array(['Premium Gasoline', 'Regular Gasoline', 'Diesel', 'Electricity',
       'Midgrade Gasoline', nan, 'Natural Gas'], dtype=object)

In [244]:
vehicle_fuel.fuelType2.unique()

array([nan, 'E85', 'Electricity', 'Natural Gas', 'Propane'], dtype=object)

### En el conjunto de datos, encontramos tres columnas que clasifican el tipo de combustible utilizado por los vehículos. Sin embargo, debido a la presencia de vehículos híbridos, la columna 'fuelType' a veces contiene un solo tipo de combustible y otras veces contiene dos tipos de combustible. Por lo tanto, hemos decidido eliminar la columna 'fuelType' y conservar únicamente las columnas 'fuelType1' y 'fuelType2'.

In [245]:
eliminar = ['barrelsA08','charge240','fuelType','barrels08','range',
       'city08U','cityA08','cityA08U','cityCD','cityE','cityUF',
       'comb08', 'comb08U','combA08', 'combA08U', 'combE', 'combinedCD', 'combinedUF',
       'cylinders','displ', 'drive', 'engId', 'eng_dscr', 'feScore', 'fuelCost08',
       'fuelCostA08', 'ghgScore', 'ghgScoreA','co2A', 'co2TailpipeAGpm', 'co2TailpipeGpm',
       'highway08', 'highway08U', 'highwayA08', 'highwayA08U',
       'highwayCD', 'highwayE', 'highwayUF', 'hlv', 'hpv', 'id', 'lv2', 'lv4',
       'mpgData', 'phevBlended', 'pv2', 'pv4','rangeCity',
       'rangeCityA', 'rangeHwy', 'rangeHwyA', 'trany', 'UCity', 'UCityA',
       'UHighway', 'UHighwayA', 'youSaveSpend', 'guzzler', 'trans_dscr',
       'tCharger', 'sCharger', 'atvType', 'evMotor','rangeA',
       'mfrCode', 'c240Dscr', 'charge240b', 'c240bDscr', 'createdOn',
       'modifiedOn', 'startStop', 'phevCity', 'phevHwy', 'phevComb']
vehicle_fuel = vehicle_fuel.drop(eliminar, axis=1)
vehicle_fuel.head(2)

Year Manufacturer         Model  city08    co2         fuelType1  \
0  2017          BMW      M6 Coupe    15.0  519.0  Premium Gasoline   
1  2017          BMW  X3 sDrive28i    21.0  374.0  Premium Gasoline   

                            VClass fuelType2  
0                  Subcompact Cars       NaN  
1  Small Sport Utility Vehicle 2WD       NaN

In [246]:
# Renombramos las columnas
vehicle_fuel = vehicle_fuel.copy()
vehicle_fuel.rename(columns={'city08':'Miles per gallon (mpg)','co2':'CO2 (p/mile)','fuelType1':'Fuel',
                             'VClass':'Category','fuelType2':'Alternative Fuel'}, inplace=True)
vehicle_fuel.head(2)

Year Manufacturer         Model  Miles per gallon (mpg)  CO2 (p/mile)  \
0  2017          BMW      M6 Coupe                    15.0         519.0   
1  2017          BMW  X3 sDrive28i                    21.0         374.0   

               Fuel                         Category Alternative Fuel  
0  Premium Gasoline                  Subcompact Cars              NaN  
1  Premium Gasoline  Small Sport Utility Vehicle 2WD              NaN

In [247]:
# Renombramos el contenido de la columna Fuel haciendo que 'Premium Gasoline', 'Regular Gasoline' y 'Midgrade Gasoline' sean 'Gasoline',
# 'Electricity' sea 'Electric' y 'Natural Gas' sea 'CNG' a efectos de normalizar contenido con otras tablas
vehicle_fuel['Fuel'] = vehicle_fuel['Fuel'].replace(['Premium Gasoline', 'Regular Gasoline', 'Midgrade Gasoline'], 'Gasoline')
vehicle_fuel['Fuel'] = vehicle_fuel['Fuel'].replace(['Electricity'], 'Electric')
vehicle_fuel['Fuel'] = vehicle_fuel['Fuel'].replace(['Natural Gas'], 'CNG')

# Renombramos el contenido de la columna Alternative Fuel haciendo que 'Electricity' sea 'Electric', 'Natural Gas' sea 'CNG' y 'Propane'
# sea 'LPG' a efectos de normalizar contenido con otras tablas
vehicle_fuel['Alternative Fuel'] = vehicle_fuel['Alternative Fuel'].replace(['Electricity'], 'Electric')
vehicle_fuel['Alternative Fuel'] = vehicle_fuel['Alternative Fuel'].replace(['Natural Gas'], 'CNG')
vehicle_fuel['Alternative Fuel'] = vehicle_fuel['Alternative Fuel'].replace(['Propane'], 'LPG')

vehicle_fuel.head(2)

Year Manufacturer         Model  Miles per gallon (mpg)  CO2 (p/mile)  \
0  2017          BMW      M6 Coupe                    15.0         519.0   
1  2017          BMW  X3 sDrive28i                    21.0         374.0   

       Fuel                         Category Alternative Fuel  
0  Gasoline                  Subcompact Cars              NaN  
1  Gasoline  Small Sport Utility Vehicle 2WD              NaN

In [248]:
# Eliminamos las filas donde 'Fuel' es nulo
vehicle_fuel = vehicle_fuel[vehicle_fuel['Fuel'].notna()]
vehicle_fuel.head(2)

Year Manufacturer         Model  Miles per gallon (mpg)  CO2 (p/mile)  \
0  2017          BMW      M6 Coupe                    15.0         519.0   
1  2017          BMW  X3 sDrive28i                    21.0         374.0   

       Fuel                         Category Alternative Fuel  
0  Gasoline                  Subcompact Cars              NaN  
1  Gasoline  Small Sport Utility Vehicle 2WD              NaN

### Dado que los campos vacíos en la columna 'Alternative Fuel' corresponden a vehículos que no son híbridos y que tienen un valor en la columna 'Fuel', hemos decidido completar los valores faltantes con 'No'..

In [249]:
# Completamos los valores nulos de la columna 'Alternative Fuel' con 'No'
vehicle_fuel['Alternative Fuel'] = vehicle_fuel['Alternative Fuel'].fillna('No')
vehicle_fuel.head(2)

Year Manufacturer         Model  Miles per gallon (mpg)  CO2 (p/mile)  \
0  2017          BMW      M6 Coupe                    15.0         519.0   
1  2017          BMW  X3 sDrive28i                    21.0         374.0   

       Fuel                         Category Alternative Fuel  
0  Gasoline                  Subcompact Cars               No  
1  Gasoline  Small Sport Utility Vehicle 2WD               No

In [250]:
# Mantenemos solo las filas donde Year es 2019 o posterior
vehicle_fuel = vehicle_fuel[vehicle_fuel['Year'] >= 2019]
vehicle_fuel.head(2)

Year Manufacturer                         Model  Miles per gallon (mpg)  \
562  2019         Jeep                  Cherokee 4WD                    19.0   
563  2019         Jeep  Cherokee 4WD Active Drive II                    18.0   

     CO2 (p/mile)      Fuel                         Category Alternative Fuel  
562         400.0  Gasoline  Small Sport Utility Vehicle 4WD               No  
563         416.0  Gasoline  Small Sport Utility Vehicle 4WD               No

In [251]:
# Revisamos la columna 'Category' y su info
vehicle_fuel['Category'].unique()

array(['Small Sport Utility Vehicle 4WD',
       'Small Sport Utility Vehicle 2WD', 'Compact Cars',
       'Small Pickup Trucks 2WD', 'Standard Pickup Trucks 4WD',
       'Two Seaters', 'Minicompact Cars', 'Subcompact Cars',
       'Midsize Cars', 'Large Cars', 'Standard Sport Utility Vehicle 4WD',
       'Small Station Wagons', 'Special Purpose Vehicle 4WD',
       'Special Purpose Vehicle 2WD', 'Midsize Station Wagons',
       'Standard Sport Utility Vehicle 2WD', 'Standard Pickup Trucks 2WD',
       'Minivan - 2WD', 'Small Pickup Trucks 4WD', 'Vans',
       'Minivan - 4WD'], dtype=object)

In [252]:
# Revisamos duplicados si hay
vehicle_fuel.duplicated().sum()

46

In [253]:
# Eliminamos las filas con duplicados
vehicle_fuel = vehicle_fuel.drop_duplicates()
vehicle_fuel.duplicated().sum()

0

In [254]:
vehicle_fuel.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6538 entries, 562 to 44949
Data columns (total 8 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Year                    6538 non-null   int64  
 1   Manufacturer            6538 non-null   object 
 2   Model                   6538 non-null   object 
 3   Miles per gallon (mpg)  6538 non-null   float64
 4   CO2 (p/mile)            6538 non-null   float64
 5   Fuel                    6538 non-null   object 
 6   Category                6538 non-null   object 
 7   Alternative Fuel        6538 non-null   object 
dtypes: float64(2), int64(1), object(5)
memory usage: 459.7+ KB


## ETL: `Air_Quality_20231216.csv`

In [255]:
# Leemos el archivo de datos
air_quality = pd.read_csv(r'C:\Users\melan\Desktop\PG 1\ETL -EDA\Data\1. Originales\Air_Quality_20231216.csv')
air_quality.head(2)

Unique ID  Indicator ID                    Name Measure Measure Info  \
0     172653           375  Nitrogen dioxide (NO2)    Mean          ppb   
1     172585           375  Nitrogen dioxide (NO2)    Mean          ppb   

  Geo Type Name  Geo Join ID                      Geo Place Name  \
0         UHF34          203  Bedford Stuyvesant - Crown Heights   
1         UHF34          203  Bedford Stuyvesant - Crown Heights   

           Time Period  Start_Date  Data Value  Message  
0  Annual Average 2011  12/01/2010       25.30      NaN  
1  Annual Average 2009  12/01/2008       26.93      NaN

In [256]:
# Revisamos el contenido del archivo
air_quality.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16218 entries, 0 to 16217
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unique ID       16218 non-null  int64  
 1   Indicator ID    16218 non-null  int64  
 2   Name            16218 non-null  object 
 3   Measure         16218 non-null  object 
 4   Measure Info    16218 non-null  object 
 5   Geo Type Name   16218 non-null  object 
 6   Geo Join ID     16218 non-null  int64  
 7   Geo Place Name  16218 non-null  object 
 8   Time Period     16218 non-null  object 
 9   Start_Date      16218 non-null  object 
 10  Data Value      16218 non-null  float64
 11  Message         0 non-null      float64
dtypes: float64(2), int64(3), object(7)
memory usage: 1.5+ MB


In [257]:
# Eliminamos las columnas que no nos interesan
eliminar = ['Unique ID', 'Indicator ID', 'Measure', 'Geo Type Name', 'Geo Join ID', 'Start_Date', 'Message']

air_quality = air_quality.drop(eliminar, axis=1)
air_quality.head(2)

Name Measure Info                      Geo Place Name  \
0  Nitrogen dioxide (NO2)          ppb  Bedford Stuyvesant - Crown Heights   
1  Nitrogen dioxide (NO2)          ppb  Bedford Stuyvesant - Crown Heights   

           Time Period  Data Value  
0  Annual Average 2011       25.30  
1  Annual Average 2009       26.93

In [258]:
# Mantenemos solo las filas donde el año es 2021
air_quality = air_quality[air_quality['Time Period'].str.contains('2021')]
air_quality.head(2)

Name Measure Info            Geo Place Name  \
12  Nitrogen dioxide (NO2)          ppb  High Bridge - Morrisania   
18  Nitrogen dioxide (NO2)          ppb                 Rockaways   

            Time Period  Data Value  
12  Annual Average 2021       18.04  
18  Annual Average 2021       11.41

In [259]:
# Mantenemos las filas donde el periodo es 'Annual Average 2021' para utilizar la media
air_quality = air_quality[air_quality['Time Period'] == 'Annual Average 2021']
air_quality['Year'] = 2023

In [260]:
# Eliminamos la columna 'Time Period' ya que no nos aporta información
air_quality = air_quality.drop('Time Period', axis=1)

In [261]:
# Define la función para asignar el borough basado en el Geo Place Name
def asignar_borough(geo_place_name):
    # Aquí colocas tu lógica para asignar el borough basado en el Geo Place Name
    # Por ejemplo:
    if "Manhattan" in geo_place_name:
        return "Manhattan"
    elif "Brooklyn" in geo_place_name:
        return "Brooklyn"
    elif "Queens" in geo_place_name:
        return "Queens"
    elif "Bronx" in geo_place_name:
        return "Bronx"
    elif "Staten Island" in geo_place_name:
        return "Staten Island"
    else:
        return "Unknown"


In [262]:
# Creamos una columna con los boroughs aplicando la función `asignar_borough`
air_quality['Borough'] = air_quality['Geo Place Name'].apply(asignar_borough)

In [263]:
# Eliminamos la columna 'Geo Place Name' ya que no es necesaria al conocer el borough
air_quality = air_quality.drop('Geo Place Name', axis=1)

In [264]:
# Aplicamos la función `borough_id` para crear la columna 'BoroughID'
air_quality['BoroughID'] = air_quality['Borough'].apply(borough_id)

In [265]:
air_quality.info()

<class 'pandas.core.frame.DataFrame'>
Index: 282 entries, 12 to 16058
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Name          282 non-null    object 
 1   Measure Info  282 non-null    object 
 2   Data Value    282 non-null    float64
 3   Year          282 non-null    int64  
 4   Borough       282 non-null    object 
 5   BoroughID     42 non-null     float64
dtypes: float64(2), int64(1), object(3)
memory usage: 15.4+ KB


In [266]:
# Agrupamos por 'Name', 'Borough', 'BoroughID' y 'Measure Info' y calculamos la media de 'Data Value'
air_quality = air_quality.groupby(['Name', 'Borough', 'BoroughID', 'Measure Info', 'Year']).agg({'Data Value': 'mean'}).reset_index()

In [267]:
# Renombramos la columna "Name" a "Pollutant"
air_quality.rename(columns={'Name': 'Pollutant'}, inplace=True)

In [268]:
# Verificamos nuestro DataFrame
air_quality

Pollutant        Borough  BoroughID Measure Info  Year  \
0  Fine particles (PM 2.5)          Bronx        2.0       mcg/m3  2023   
1  Fine particles (PM 2.5)       Brooklyn        3.0       mcg/m3  2023   
2  Fine particles (PM 2.5)      Manhattan        1.0       mcg/m3  2023   
3  Fine particles (PM 2.5)         Queens        4.0       mcg/m3  2023   
4  Fine particles (PM 2.5)  Staten Island        5.0       mcg/m3  2023   
5   Nitrogen dioxide (NO2)          Bronx        2.0          ppb  2023   
6   Nitrogen dioxide (NO2)       Brooklyn        3.0          ppb  2023   
7   Nitrogen dioxide (NO2)      Manhattan        1.0          ppb  2023   
8   Nitrogen dioxide (NO2)         Queens        4.0          ppb  2023   
9   Nitrogen dioxide (NO2)  Staten Island        5.0          ppb  2023   

   Data Value  
0    6.594286  
1    6.865000  
2    7.573333  
3    6.520000  
4    6.150000  
5   16.424286  
6   17.740000  
7   20.540000  
8   15.737500  
9   12.840000

### Guardamos los archivos en formato parquet

In [269]:
air_quality.to_parquet(r'C:\Users\melan\Desktop\PG 1\ETL -EDA\Data\2. Normalizacion/air_quality.parquet', index=False)

## ETL: `Alternative Fuel Vehicles US.csv`

In [270]:
# Leemos el archivo de datos 
a_fuel_veh_US = pd.read_csv(r'C:\Users\melan\Desktop\PG 1\ETL -EDA\Data\1. Originales\Alternative Fuel Vehicles US.csv')
a_fuel_veh_US.head(2)

Category Model  Model Year Manufacturer             Fuel  \
0  Sedan/Wagon   NSX      2022.0        Acura  Hybrid Electric   
1  Sedan/Wagon   A3       2022.0         Audi  Hybrid Electric   

   All-Electric Range  PHEV Total Range  Alternative Fuel Economy City  \
0                 NaN               NaN                            NaN   
1                 NaN               NaN                            NaN   

   Alternative Fuel Economy Highway  Alternative Fuel Economy Combined  ...  \
0                               NaN                                NaN  ...   
1                               NaN                                NaN  ...   

   Conventional Fuel Economy Combined  Transmission Type  Transmission Make  \
0                                21.0               Auto                NaN   
1                                32.0               Auto                NaN   

  Engine Type Engine Size Engine Cylinder Count Number of Passengers  \
0          SI        3.5L                   6.0                  NaN   
1          SI        2.0L                   4.0                  NaN   

   Heavy-Duty Power System Notes Drivetrain  
0                      NaN   NaN        AWD  
1                      NaN   NaN        FWD  

[2 rows x 22 columns]

In [271]:
a_fuel_veh_US['Fuel'].unique()

array(['Hybrid Electric', 'Plug-in Hybrid Electric', 'Electric',
       'Biodiesel (B20)', 'Ethanol (E85)', 'Hydrogen Fuel Cell',
       'CNG - Bi-fuel', 'CNG - Compressed Natural Gas',
       'Propane - Bi-fuel', 'Propane',
       'LNG - Liquified Natural Gas|CNG - Compressed Natural Gas',
       'CNG - Compressed Natural Gas|LNG - Liquified Natural Gas',
       'Propane|CNG - Compressed Natural Gas',
       'CNG - Compressed Natural Gas|Plug-in Hybrid Electric',
       'Diesel/Hybrid Electric',
       'Propane - Bi-fuel|CNG - Bi-fuel|Propane|CNG - Compressed Natural Gas|Ethanol (E85)',
       'Propane - Bi-fuel|Propane|CNG - Bi-fuel|Ethanol (E85)|CNG - Compressed Natural Gas',
       'CNG - Compressed Natural Gas|Propane - Bi-fuel|E85/Hybrid Electric|CNG - Bi-fuel|Propane',
       'CNG - Bi-fuel|Ethanol (E85)|CNG - Compressed Natural Gas|Propane|Propane - Bi-fuel',
       'Propane - Bi-fuel|Propane|CNG - Compressed Natural Gas|CNG - Bi-fuel',
       'CNG - Compressed Natural Gas|Prop

In [272]:
a_fuel_veh_US['Conventional Fuel Economy City'].unique()

array([21., 29., 28., 24., 26., 23., 25., 19., nan, 18., 15., 13., 14.,
       17., 22., 20., 16., 43., 44., 27., 42., 48., 40., 55., 51., 49.,
       53., 54., 58., 52., 33., 36., 45., 50., 37., 38., 35., 46., 39.,
       41., 31., 34., 30., 59.])

In [273]:
a_fuel_veh_US['Alternative Fuel Economy City'].unique()

array([ nan,  81.,  78.,  72.,  62.,  76.,  73.,  64., 100., 109.,  94.,
        79.,  86.,  82., 125., 131.,  11.,  10.,  14.,  12.,  98., 105.,
        97.,  90.,  88., 110., 108., 104.,  18.,  16., 127., 132.,  59.,
        65., 116., 136., 134., 123., 117., 126., 115., 130., 121.,  92.,
       119., 118., 114., 113.,  75.,  71.,  69.,  74., 138., 124., 102.,
       107., 103., 129., 140.,  67., 106.,  96.,  17.,  68., 145.,  80.,
       141., 150.,  85.])

In [274]:
# Revisamos el contenido del archivo
a_fuel_veh_US.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 882 entries, 0 to 881
Data columns (total 22 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Category                            882 non-null    object 
 1   Model                               882 non-null    object 
 2   Model Year                          594 non-null    float64
 3   Manufacturer                        882 non-null    object 
 4   Fuel                                882 non-null    object 
 5   All-Electric Range                  222 non-null    float64
 6   PHEV Total Range                    43 non-null     float64
 7   Alternative Fuel Economy City       169 non-null    float64
 8   Alternative Fuel Economy Highway    169 non-null    float64
 9   Alternative Fuel Economy Combined   262 non-null    float64
 10  Conventional Fuel Economy City      444 non-null    float64
 11  Conventional Fuel Economy Highway   444 non-n

In [275]:
# Eliminamos las columnas que no nos interesan
eliminar = ['All-Electric Range','PHEV Total Range','Alternative Fuel Economy City','Alternative Fuel Economy Highway',
            'Alternative Fuel Economy Combined','Conventional Fuel Economy City','Conventional Fuel Economy Highway',
            'Conventional Fuel Economy Combined','Transmission Type','Transmission Make','Engine Type','Engine Size',
            'Engine Cylinder Count','Number of Passengers','Heavy-Duty Power System','Notes','Drivetrain']

a_fuel_veh_US = a_fuel_veh_US.drop(eliminar, axis=1)
a_fuel_veh_US.head(2)

Category Model  Model Year Manufacturer             Fuel
0  Sedan/Wagon   NSX      2022.0        Acura  Hybrid Electric
1  Sedan/Wagon   A3       2022.0         Audi  Hybrid Electric

In [276]:
# Revisamos el contenido de la columna 'Category'
a_fuel_veh_US.Category.unique()

array(['Sedan/Wagon', 'SUV', 'Pickup', 'Van', 'Passenger Van/Shuttle Bus',
       'Street Sweeper', 'Refuse', 'Vocational/Cab Chassis', 'Tractor',
       'School Bus', 'Transit Bus', 'Step Van',
       'TractorVocational/Cab Chassis', 'Vocational/Cab ChassisVan',
       'Vocational/Cab ChassisTractor'], dtype=object)

#### Como en NYC los vehículos habilitados como Taxi son automóviles sedan o SUV, mantenemos solo los registros de estos tipos de vehículos.

In [277]:
# Se mantiene solo las filas donde la categoría es 'Sedan/Wagon' o 'SUV'
a_fuel_veh_US = a_fuel_veh_US[(a_fuel_veh_US['Category'] == 'Sedan/Wagon') | (a_fuel_veh_US['Category'] == 'SUV')]
a_fuel_veh_US.Category.unique()

array(['Sedan/Wagon', 'SUV'], dtype=object)

In [278]:
a_fuel_veh_US.info()

<class 'pandas.core.frame.DataFrame'>
Index: 505 entries, 0 to 593
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Category      505 non-null    object 
 1   Model         505 non-null    object 
 2   Model Year    505 non-null    float64
 3   Manufacturer  505 non-null    object 
 4   Fuel          505 non-null    object 
dtypes: float64(1), object(4)
memory usage: 23.7+ KB


In [279]:
# Convertimos la columna 'Model Year' a tipo int
a_fuel_veh_US['Model Year'] = a_fuel_veh_US['Model Year'].astype(int)
a_fuel_veh_US['Model Year'].unique()

array([2022, 2021])

## ETL: `annotations.csv`

In [280]:
# Cargamos el archivo de datos
sound_quality = pd.read_csv('Datasets/1-Provistos/annotations.csv')
sound_quality.head(2)

FileNotFoundError: [Errno 2] No such file or directory: 'Datasets/1-Provistos/annotations.csv'

In [ ]:
# Eliminamos las columnas que no nos interesan
eliminar = ['split','sensor_id','audio_filename','annotator_id','block','1-1_small-sounding-engine_presence','week',
            '1-2_medium-sounding-engine_presence','1-3_large-sounding-engine_presence','1-X_engine-of-uncertain-size_presence',
            '2-1_rock-drill_presence','2-2_jackhammer_presence','2-3_hoe-ram_presence','2-4_pile-driver_presence',
            '2-X_other-unknown-impact-machinery_presence','3-1_non-machinery-impact_presence','4-1_chainsaw_presence',
            '4-2_small-medium-rotating-saw_presence','4-3_large-rotating-saw_presence','4-X_other-unknown-powered-saw_presence',
            '5-1_car-horn_presence','5-2_car-alarm_presence','5-3_siren_presence','5-4_reverse-beeper_presence',
            '5-X_other-unknown-alert-signal_presence','6-1_stationary-music_presence','6-2_mobile-music_presence',
            '6-3_ice-cream-truck_presence','6-X_music-from-uncertain-source_presence','7-1_person-or-small-group-talking_presence',
            '7-2_person-or-small-group-shouting_presence','7-3_large-crowd_presence','7-4_amplified-speech_presence',
            '7-X_other-unknown-human-voice_presence','8-1_dog-barking-whining_presence','1-1_small-sounding-engine_proximity',
            '1-2_medium-sounding-engine_proximity','1-3_large-sounding-engine_proximity','1-X_engine-of-uncertain-size_proximity',
            '2-1_rock-drill_proximity','2-2_jackhammer_proximity','2-3_hoe-ram_proximity','2-4_pile-driver_proximity',
            '2-X_other-unknown-impact-machinery_proximity','3-1_non-machinery-impact_proximity','4-1_chainsaw_proximity',
            '4-2_small-medium-rotating-saw_proximity','4-3_large-rotating-saw_proximity','4-X_other-unknown-powered-saw_proximity',
            '5-1_car-horn_proximity','5-2_car-alarm_proximity','5-3_siren_proximity','5-4_reverse-beeper_proximity',
            '5-X_other-unknown-alert-signal_proximity','6-1_stationary-music_proximity','6-2_mobile-music_proximity',
            '6-3_ice-cream-truck_proximity','6-X_music-from-uncertain-source_proximity','7-1_person-or-small-group-talking_proximity',
            '7-2_person-or-small-group-shouting_proximity','7-3_large-crowd_proximity','7-4_amplified-speech_proximity',
            '7-X_other-unknown-human-voice_proximity','8-1_dog-barking-whining_proximity','2_machinery-impact_presence',
            '3_non-machinery-impact_presence','4_powered-saw_presence','5_alert-signal_presence','6_music_presence','7_human-voice_presence',
            '8_dog_presence']

sound_quality = sound_quality.drop(eliminar, axis=1)
sound_quality

borough  latitude  longitude  year  day  hour  1_engine_presence
0            1  40.72951  -73.99388  2019    4     8                 -1
1            1  40.72951  -73.99388  2019    4     8                  1
2            1  40.72951  -73.99388  2019    4     8                  1
3            1  40.72951  -73.99388  2019    4     8                 -1
4            1  40.72951  -73.99388  2019    4     8                  1
...        ...       ...        ...   ...  ...   ...                ...
62017        1  40.73365  -73.98879  2018    5    19                  0
62018        1  40.73365  -73.98879  2018    5    19                  0
62019        1  40.73365  -73.98879  2018    2     9                  0
62020        1  40.73365  -73.98879  2018    2     9                  0
62021        1  40.73365  -73.98879  2018    2     9                  1

[62022 rows x 7 columns]

In [ ]:
# Mantenemos solo las filas donde year es 2019
sound_quality = sound_quality[sound_quality['year'] == 2019]
sound_quality

borough  latitude  longitude  year  day  hour  1_engine_presence
0            1  40.72951  -73.99388  2019    4     8                 -1
1            1  40.72951  -73.99388  2019    4     8                  1
2            1  40.72951  -73.99388  2019    4     8                  1
3            1  40.72951  -73.99388  2019    4     8                 -1
4            1  40.72951  -73.99388  2019    4     8                  1
...        ...       ...        ...   ...  ...   ...                ...
61993        1  40.73365  -73.98879  2019    1    20                  0
61994        1  40.73365  -73.98879  2019    1    20                  0
62007        1  40.73365  -73.98879  2019    3    10                  0
62008        1  40.73365  -73.98879  2019    3    10                  0
62009        1  40.73365  -73.98879  2019    3    10                  0

[8840 rows x 7 columns]

In [ ]:
# Buscamos filas repetidas
sound_quality.duplicated().sum()

6078

In [ ]:
# Eliminamos las filas repetidas
sound_quality = sound_quality.drop_duplicates()

In [ ]:
# Aplicamos la función `engine_sound` para crear la columna 'Engine Sound'
sound_quality = sound_quality.copy()
sound_quality['engine_sound'] = sound_quality['1_engine_presence'].apply(engine_sound)

In [ ]:
# Eliminamos la columna '1_engine_presence' ya que no es necesaria al conocer el nivel de sonido del motor
sound_quality = sound_quality.drop('1_engine_presence', axis=1)
sound_quality.head(2)

borough  latitude  longitude  year  day  hour engine_sound
0        1  40.72951  -73.99388  2019    4     8          Low
1        1  40.72951  -73.99388  2019    4     8         High

In [ ]:
# Renombramos las columnas y reseteamos el índice
sound_quality.rename(columns={'borough':'Borough','latitude':'Latitude','longitude':'Longitude','year':'Year',
                              'day':'Day','hour':'Hour','engine_sound':'Engine Sound'}, inplace=True)

sound_quality.reset_index(drop=True)

Borough  Latitude  Longitude  Year  Day  Hour Engine Sound
0           1  40.72951  -73.99388  2019    4     8          Low
1           1  40.72951  -73.99388  2019    4     8         High
2           1  40.72951  -73.99388  2019    1    20          Low
3           1  40.72951  -73.99388  2019    1    20         High
4           1  40.72951  -73.99388  2019    1    20       Medium
...       ...       ...        ...   ...  ...   ...          ...
2757        1  40.73365  -73.98879  2019    4     8         High
2758        1  40.73365  -73.98879  2019    0     7       Medium
2759        1  40.73365  -73.98879  2019    4    16         High
2760        1  40.73365  -73.98879  2019    6     8         High
2761        1  40.73365  -73.98879  2019    3    10       Medium

[2762 rows x 7 columns]

### Guardamos el archivo en formato parquet

In [ ]:
sound_quality.to_parquet('Datasets/3-Normalizados/sound_quality.parquet', index=False)

## ETL: `MY2012-2023_Battery_Electric_Vehicles.csv`

In [ ]:
# Leemos el archivo de datos
bat_elec_veh = pd.read_csv(r'C:\Users\melan\Desktop\PG 1\ETL -EDA\Data\1. Originales\MY2012-2023_Battery_Electric_Vehicles.csv', encoding='latin-1')
bat_elec_veh.head(2)

Model        Make Model.1 Vehicle Class Motor Transmission  Fuel  \
0  Year         NaN     NaN           NaN  (kW)          NaN  Type   
1  2012  Mitsubishi  i-MiEV    Subcompact    49           A1     B   

         Consumption        Unnamed: 8         Unnamed: 9       Unnamed: 10  \
0  City (kWh/100 km)  Hwy (kWh/100 km)  Comb (kWh/100 km)  City (Le/100 km)   
1               16.9              21.4               18.7               1.9   

       Unnamed: 11       Unnamed: 12 Range CO2 Emissions    CO2     Smog  \
0  Hwy (Le/100 km)  Comb (Le/100 km)  (km)        (g/km)  Rating  Rating   
1              2.4               2.1   100             0     NaN     NaN   

   Recharge  
0  Time (h)  
1         7

In [ ]:
# Revisamos la composición del archivo
bat_elec_veh.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 486 entries, 0 to 485
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Model          482 non-null    object
 1   Make           480 non-null    object
 2   Model.1        470 non-null    object
 3   Vehicle Class  470 non-null    object
 4   Motor          471 non-null    object
 5   Transmission   470 non-null    object
 6   Fuel           471 non-null    object
 7   Consumption    471 non-null    object
 8   Unnamed: 8     471 non-null    object
 9   Unnamed: 9     471 non-null    object
 10  Unnamed: 10    471 non-null    object
 11  Unnamed: 11    471 non-null    object
 12  Unnamed: 12    471 non-null    object
 13  Range          471 non-null    object
 14  CO2 Emissions  471 non-null    object
 15  CO2            437 non-null    object
 16  Smog           410 non-null    object
 17  Recharge       471 non-null    object
dtypes: object(18)
memory usage: 68

In [ ]:
# Eliminamos las columnas que no nos interesan
eliminar = ['Motor','Transmission','Unnamed: 8','Unnamed: 9','Unnamed: 10',
            'Unnamed: 11','Unnamed: 12','Range','CO2 ','Smog','Recharge']

bat_elec_veh = bat_elec_veh.drop(eliminar, axis=1)

In [ ]:
bat_elec_veh.head(2)

Model        Make Model.1 Vehicle Class  Fuel        Consumption  \
0  Year         NaN     NaN           NaN  Type  City (kWh/100 km)   
1  2012  Mitsubishi  i-MiEV    Subcompact     B               16.9   

  CO2 Emissions  
0        (g/km)  
1             0

In [ ]:
# Renombramos las columnas
bat_elec_veh.rename(columns={'Model':'Year','Make':'Manufacturer','Model.1':'Model','Consumption':'Consumption',
                             'Range':'Range (miles)','CO2 Emissions':'CO2 (mg/mile)','Recharge':'Recharge (hours)'}, inplace=True)

In [ ]:
bat_elec_veh.head(2)

Year Manufacturer   Model Vehicle Class  Fuel        Consumption  \
0  Year          NaN     NaN           NaN  Type  City (kWh/100 km)   
1  2012   Mitsubishi  i-MiEV    Subcompact     B               16.9   

  CO2 (mg/mile)  
0        (g/km)  
1             0

In [ ]:
# Eliminamos la segunda fila del dataframe
bat_elec_veh = bat_elec_veh.drop(0, axis=0)
bat_elec_veh.head(2)

Year Manufacturer   Model Vehicle Class Fuel Consumption CO2 (mg/mile)
1  2012   Mitsubishi  i-MiEV    Subcompact    B        16.9             0
2  2012       Nissan    LEAF      Mid-size    B        19.3             0

In [ ]:
# Define la función para editar la columna 'Fuel'
def fuel(fuel_value):
    # Aquí colocas tu lógica para editar los valores de la columna 'Fuel'
    # Por ejemplo, puedes reemplazar ciertos valores, eliminar caracteres no deseados, etc.
    # Aquí proporciono un ejemplo genérico:
    if fuel_value == 'Gasoline':
        return 'Gasolina'
    elif fuel_value == 'Electric':
        return 'Eléctrico'
    else:
        return fuel_value  # Devuelve el valor original si no se requiere ninguna edición


In [ ]:
# Aplicamos la función `fuel` para editar la columna 'Fuel'
bat_elec_veh['Fuel'] = bat_elec_veh['Fuel'].apply(fuel)
bat_elec_veh.head(2)

Year Manufacturer   Model Vehicle Class Fuel Consumption CO2 (mg/mile)
1  2012   Mitsubishi  i-MiEV    Subcompact    B        16.9             0
2  2012       Nissan    LEAF      Mid-size    B        19.3             0

In [ ]:
bat_elec_veh['Vehicle Class'].unique()

array(['Subcompact', 'Mid-size', 'Compact', 'Two-seater', 'Full-size',
       'Station wagon: Small', 'SUV: Standard', 'SUV: Small',
       'Pickup truck: Standard', nan], dtype=object)

In [ ]:
# Mantenemos solo las filas donde 'Vehicle Class' es 'Full-size', 'SUV: Small', 'SUV: Standard', 'Compact', 'Mid-size'
bat_elec_veh = bat_elec_veh[(bat_elec_veh['Vehicle Class'] == 'Full-size') | (bat_elec_veh['Vehicle Class'] == 'SUV: Small') |
                            (bat_elec_veh['Vehicle Class'] == 'SUV: Standard') | (bat_elec_veh['Vehicle Class'] == 'Compact') |
                            (bat_elec_veh['Vehicle Class'] == 'Mid-size')]

bat_elec_veh['Vehicle Class'].unique()

array(['Mid-size', 'Compact', 'Full-size', 'SUV: Standard', 'SUV: Small'],
      dtype=object)

In [ ]:
# Verificamos si hay duplicados en la columna 'Model'
bat_elec_veh['Model'].duplicated().sum()

167

In [ ]:
# Eliminamos las filas duplicadas de la columna 'Model'
bat_elec_veh = bat_elec_veh.drop_duplicates(subset=['Model'])
bat_elec_veh

Year Manufacturer                 Model Vehicle Class Fuel Consumption  \
2    2012       Nissan                  LEAF      Mid-size    B        19.3   
3    2013         Ford        Focus Electric       Compact    B        19.0   
8    2013        Tesla      Model S (40 kWh)     Full-size    B        22.4   
9    2013        Tesla      Model S (60 kWh)     Full-size    B        22.2   
10   2013        Tesla      Model S (85 kWh)     Full-size    B        23.8   
..    ...          ...                   ...           ...  ...         ...   
462  2023       Toyota              bZ4X AWD    SUV: Small    B        18.4   
463  2023      VinFast               VF8 ECO    SUV: Small    B        26.4   
464  2023      VinFast              VF8 PLUS    SUV: Small    B        32.3   
465  2023      VinFast  VF8 PLUS Performance    SUV: Small    B        43.8   
466  2023   Volkswagen                  ID.4    SUV: Small    B        18.2   

    CO2 (mg/mile)  
2               0  
3               0  
8               0  
9               0  
10              0  
..            ...  
462             0  
463             0  
464             0  
465             0  
466             0  

[192 rows x 7 columns]

## ETL: `MY2023 Fuel Consumption Ratings.csv`

In [ ]:
# Leemos el archivo de datos
fuel_cons_rat = pd.read_csv(r'C:\Users\melan\Desktop\PG 1\ETL -EDA\Data\1. Originales\MY2023 Fuel Consumption Ratings.csv', encoding='latin-1')
fuel_cons_rat.head(2)

Model   Make  Model.1 Vehicle Class Engine Size  Cylinders Transmission  \
0  Year    NaN      NaN           NaN         (L)        NaN          NaN   
1  2023  Acura  Integra     Full-size         1.5        4.0          AV7   

   Fuel Fuel Consumption      Unnamed: 9  ... Unnamed: 214 Unnamed: 215  \
0  Type  City (L/100 km)  Hwy (L/100 km)  ...          NaN          NaN   
1     Z              7.9             6.3  ...          NaN          NaN   

  Unnamed: 216 Unnamed: 217 Unnamed: 218  Unnamed: 219  Unnamed: 220  \
0          NaN          NaN          NaN           NaN           NaN   
1          NaN          NaN          NaN           NaN           NaN   

   Unnamed: 221  Unnamed: 222  Unnamed: 223  
0           NaN           NaN           NaN  
1           NaN           NaN           NaN  

[2 rows x 224 columns]

### Observamos que de 224 columnas solo 15 tienen datos, por lo que eliminamos las columnas que no tienen datos.

In [ ]:
# Mantenemos solo las columnas que nos interesan
columns_to_keep = ['Model', 'Make', 'Model.1', 'Vehicle Class', 'Engine Size', 'Cylinders',
                   'Transmission', 'Fuel', 'Fuel Consumption', 'Unnamed: 9', 'Unnamed: 10',
                   'Unnamed: 11', 'CO2 Emissions', 'CO2', 'Smog']

fuel_cons_rat = fuel_cons_rat[columns_to_keep]
fuel_cons_rat.head(2)

Model   Make  Model.1 Vehicle Class Engine Size  Cylinders Transmission  \
0  Year    NaN      NaN           NaN         (L)        NaN          NaN   
1  2023  Acura  Integra     Full-size         1.5        4.0          AV7   

   Fuel Fuel Consumption      Unnamed: 9      Unnamed: 10 Unnamed: 11  \
0  Type  City (L/100 km)  Hwy (L/100 km)  Comb (L/100 km)  Comb (mpg)   
1     Z              7.9             6.3              7.2          39   

  CO2 Emissions     CO2    Smog  
0        (g/km)  Rating  Rating  
1           167       6       7

In [ ]:
# Renombramos las columnas
fuel_cons_rat.rename(columns={'Model':'Year', 'Make':'Manufacturer','Model.1':'Model', 'Fuel Consumption':'Consumption',
                              'Unnamed: 9':'Hwy (G/100 km)', 'Unnamed: 10':'Comb (G/100 km)',
                              'Unnamed: 11':'Comb (mpg)','CO2 Emissions':'CO2 (mg/mile)'}, inplace=True)

In [ ]:
fuel_cons_rat.head(2)

Year Manufacturer    Model Vehicle Class Engine Size  Cylinders  \
0  Year          NaN      NaN           NaN         (L)        NaN   
1  2023        Acura  Integra     Full-size         1.5        4.0   

  Transmission  Fuel      Consumption  Hwy (G/100 km)  Comb (G/100 km)  \
0          NaN  Type  City (L/100 km)  Hwy (L/100 km)  Comb (L/100 km)   
1          AV7     Z              7.9             6.3              7.2   

   Comb (mpg) CO2 (mg/mile)     CO2    Smog  
0  Comb (mpg)        (g/km)  Rating  Rating  
1          39           167       6       7

In [ ]:
# Elegimos columnas a eliminar dentro de las sobrantes
eliminar = ['Engine Size', 'Cylinders', 'Transmission', 'Transmission', 'Hwy (G/100 km)', 'Comb (G/100 km)',
            'Comb (mpg)', 'CO2', 'Smog']

fuel_cons_rat = fuel_cons_rat.drop(eliminar, axis=1)
fuel_cons_rat.head(2)

Year Manufacturer    Model Vehicle Class  Fuel      Consumption  \
0  Year          NaN      NaN           NaN  Type  City (L/100 km)   
1  2023        Acura  Integra     Full-size     Z              7.9   

  CO2 (mg/mile)  
0        (g/km)  
1           167

In [ ]:
# Eliminamos la segunda fila del DataFrame y reseteamos el índice
fuel_cons_rat = fuel_cons_rat.drop(0, axis=0)
fuel_cons_rat.reset_index(drop=True)
fuel_cons_rat.head(2)

Year Manufacturer           Model Vehicle Class Fuel Consumption  \
1  2023        Acura         Integra     Full-size    Z         7.9   
2  2023        Acura  Integra A-SPEC     Full-size    Z         8.1   

  CO2 (mg/mile)  
1           167  
2           172

In [ ]:
# Define la función para convertir litros a galones
def convert_l_to_gal(liters):
    # 1 litro equivale a 0.264172 galones
    gallons = liters * 0.264172
    return gallons


In [ ]:
# Convertimos la columna 'Consumption' a tipo float
fuel_cons_rat['Consumption'] = fuel_cons_rat['Consumption'].astype(float)

# Aplicacamos la función `convert_l_to_gal` a la columna 'Consumption'
fuel_cons_rat['Consumption'] = fuel_cons_rat['Consumption'].apply(convert_l_to_gal)
fuel_cons_rat.head(2)


Year Manufacturer           Model Vehicle Class Fuel  Consumption  \
1  2023        Acura         Integra     Full-size    Z     2.086959   
2  2023        Acura  Integra A-SPEC     Full-size    Z     2.139793   

  CO2 (mg/mile)  
1           167  
2           172

In [ ]:
# Convertimos la columna 'CO2 (mg/mile)' a tipo float
fuel_cons_rat['CO2 (mg/mile)'] = fuel_cons_rat['CO2 (mg/mile)'].astype(float)

# Aplicamos la función `convert_km_to_mile` a la columna 'CO2 (mg/km)'
fuel_cons_rat['CO2 (mg/mile)'] = fuel_cons_rat['CO2 (mg/mile)'].apply(convert_km_to_mile)
fuel_cons_rat.head(2)

Year Manufacturer           Model Vehicle Class Fuel  Consumption  \
1  2023        Acura         Integra     Full-size    Z     2.086959   
2  2023        Acura  Integra A-SPEC     Full-size    Z     2.139793   

   CO2 (mg/mile)  
1     103.768957  
2     106.875812

In [ ]:
# Revisamos el contenido de la columna 'Vehicle Class'
fuel_cons_rat['Vehicle Class'].unique()

array(['Full-size', 'SUV: Small', 'SUV: Standard', 'Compact', 'Mid-size',
       'Minicompact', 'Two-seater', 'Subcompact', 'Station wagon: Small',
       'Station wagon: Mid-size', 'Pickup truck: Small',
       'Pickup truck: Standard', 'Minivan', 'Special purpose vehicle',
       nan], dtype=object)

In [ ]:
# Mantenemos solo las filas donde 'Vehicle Class' es 'Full-size', 'SUV: Small', 'SUV: Standard', 'Compact', 'Mid-size'
fuel_cons_rat = fuel_cons_rat[(fuel_cons_rat['Vehicle Class'] == 'Full-size') | (fuel_cons_rat['Vehicle Class'] == 'SUV: Small') |
                                (fuel_cons_rat['Vehicle Class'] == 'SUV: Standard') | (fuel_cons_rat['Vehicle Class'] == 'Compact') |
                                (fuel_cons_rat['Vehicle Class'] == 'Mid-size')]

fuel_cons_rat['Vehicle Class'].unique()

array(['Full-size', 'SUV: Small', 'SUV: Standard', 'Compact', 'Mid-size'],
      dtype=object)

In [ ]:
fuel_cons_rat

Year Manufacturer              Model  Vehicle Class Fuel  Consumption  \
1    2023        Acura            Integra      Full-size    Z     2.086959   
2    2023        Acura     Integra A-SPEC      Full-size    Z     2.139793   
3    2023        Acura     Integra A-SPEC      Full-size    Z     2.351131   
4    2023        Acura         MDX SH-AWD     SUV: Small    Z     3.328567   
5    2023        Acura  MDX SH-AWD Type S  SUV: Standard    Z     3.645574   
..    ...          ...                ...            ...  ...          ...   
829  2023        Volvo        XC40 B5 AWD     SUV: Small    Z     2.694554   
830  2023        Volvo        XC60 B5 AWD     SUV: Small    Z     2.720972   
831  2023        Volvo        XC60 B6 AWD     SUV: Small    Z     2.932309   
832  2023        Volvo        XC90 B5 AWD  SUV: Standard    Z     2.773806   
833  2023        Volvo        XC90 B6 AWD  SUV: Standard    Z     3.143647   

     CO2 (mg/mile)  
1       103.768957  
2       106.875812  
3       112.468151  
4       163.420573  
5       180.818961  
..             ...  
829     133.594765  
830     135.458878  
831     144.779443  
832     138.565733  
833     154.721379  

[552 rows x 7 columns]

In [ ]:
# Verificamos si hay duplicados en la columna 'Model'
fuel_cons_rat['Model'].duplicated().sum()

103

In [ ]:
# Eliminamos las filas duplicadas de la columna 'Model'
fuel_cons_rat = fuel_cons_rat.drop_duplicates(subset=['Model'])
fuel_cons_rat.reset_index(drop=True)
fuel_cons_rat

Year Manufacturer              Model  Vehicle Class Fuel  Consumption  \
1    2023        Acura            Integra      Full-size    Z     2.086959   
2    2023        Acura     Integra A-SPEC      Full-size    Z     2.139793   
4    2023        Acura         MDX SH-AWD     SUV: Small    Z     3.328567   
5    2023        Acura  MDX SH-AWD Type S  SUV: Standard    Z     3.645574   
6    2023        Acura         RDX SH-AWD     SUV: Small    Z     2.905892   
..    ...          ...                ...            ...  ...          ...   
829  2023        Volvo        XC40 B5 AWD     SUV: Small    Z     2.694554   
830  2023        Volvo        XC60 B5 AWD     SUV: Small    Z     2.720972   
831  2023        Volvo        XC60 B6 AWD     SUV: Small    Z     2.932309   
832  2023        Volvo        XC90 B5 AWD  SUV: Standard    Z     2.773806   
833  2023        Volvo        XC90 B6 AWD  SUV: Standard    Z     3.143647   

     CO2 (mg/mile)  
1       103.768957  
2       106.875812  
4       163.420573  
5       180.818961  
6       144.158072  
..             ...  
829     133.594765  
830     135.458878  
831     144.779443  
832     138.565733  
833     154.721379  

[449 rows x 7 columns]

In [ ]:
# Verificamos el contenido de la columna 'Fuel'
fuel_cons_rat.Fuel.unique()

array(['Z', 'X', 'D'], dtype=object)

In [ ]:
# Define la función para reemplazar valores en la columna 'Fuel'
def fuel_type(fuel_value):
    # Aquí colocas tu lógica para reemplazar los valores en la columna 'Fuel'
    # Por ejemplo, puedes asignar un valor diferente a cada tipo de combustible
    # Aquí proporciono un ejemplo genérico:
    if fuel_value == 'Gasolina':
        return 'Gasolina'
    elif fuel_value == 'Diesel':
        return 'Diesel'
    elif fuel_value == 'Eléctrico':
        return 'Eléctrico'
    else:
        return 'Otros'  # Puedes manejar valores no especificados según sea necesario


In [ ]:
# Aplicamos la función `fuel_type` reemplazando los valores en la columna 'Fuel'
fuel_cons_rat.loc[:, 'Fuel'] = fuel_cons_rat['Fuel'].apply(fuel_type)


In [ ]:
fuel_cons_rat

Year Manufacturer              Model  Vehicle Class   Fuel  Consumption  \
1    2023        Acura            Integra      Full-size  Otros     2.086959   
2    2023        Acura     Integra A-SPEC      Full-size  Otros     2.139793   
4    2023        Acura         MDX SH-AWD     SUV: Small  Otros     3.328567   
5    2023        Acura  MDX SH-AWD Type S  SUV: Standard  Otros     3.645574   
6    2023        Acura         RDX SH-AWD     SUV: Small  Otros     2.905892   
..    ...          ...                ...            ...    ...          ...   
829  2023        Volvo        XC40 B5 AWD     SUV: Small  Otros     2.694554   
830  2023        Volvo        XC60 B5 AWD     SUV: Small  Otros     2.720972   
831  2023        Volvo        XC60 B6 AWD     SUV: Small  Otros     2.932309   
832  2023        Volvo        XC90 B5 AWD  SUV: Standard  Otros     2.773806   
833  2023        Volvo        XC90 B6 AWD  SUV: Standard  Otros     3.143647   

     CO2 (mg/mile)  
1       103.768957  
2       106.875812  
4       163.420573  
5       180.818961  
6       144.158072  
..             ...  
829     133.594765  
830     135.458878  
831     144.779443  
832     138.565733  
833     154.721379  

[449 rows x 7 columns]

In [ ]:
bat_elec_veh.info()

<class 'pandas.core.frame.DataFrame'>
Index: 192 entries, 2 to 466
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Year           192 non-null    object
 1   Manufacturer   192 non-null    object
 2   Model          192 non-null    object
 3   Vehicle Class  192 non-null    object
 4   Fuel           192 non-null    object
 5   Consumption    192 non-null    object
 6   CO2 (mg/mile)  192 non-null    object
dtypes: object(7)
memory usage: 12.0+ KB


In [ ]:
fuel_cons_rat.info()

<class 'pandas.core.frame.DataFrame'>
Index: 449 entries, 1 to 833
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Year           449 non-null    object 
 1   Manufacturer   449 non-null    object 
 2   Model          449 non-null    object 
 3   Vehicle Class  449 non-null    object 
 4   Fuel           449 non-null    object 
 5   Consumption    449 non-null    float64
 6   CO2 (mg/mile)  449 non-null    float64
dtypes: float64(2), object(5)
memory usage: 28.1+ KB


### Teniendo en cuenta que los datasets electric_vehicle y fuel_consumption_ratings contienen información de vehículos y una misma estructura de columnas, los unimos en un solo dataset.

In [ ]:
# Unimos los Dataframes electric_vehicles y fuel_consumption_ratings
vehicle_consumption = pd.concat([fuel_cons_rat, bat_elec_veh], axis=0)
vehicle_consumption.reset_index(drop=True)
vehicle_consumption.info()

<class 'pandas.core.frame.DataFrame'>
Index: 641 entries, 1 to 466
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Year           641 non-null    object
 1   Manufacturer   641 non-null    object
 2   Model          641 non-null    object
 3   Vehicle Class  641 non-null    object
 4   Fuel           641 non-null    object
 5   Consumption    641 non-null    object
 6   CO2 (mg/mile)  641 non-null    object
dtypes: object(7)
memory usage: 40.1+ KB


In [ ]:
# Convertimos las columnas 'Consumption' y 'CO2 (mg/mile)' a tipo float
vehicle_consumption['Consumption'] = vehicle_consumption['Consumption'].astype(float)
vehicle_consumption['CO2 (mg/mile)'] = vehicle_consumption['CO2 (mg/mile)'].astype(float)
vehicle_consumption.info()

<class 'pandas.core.frame.DataFrame'>
Index: 641 entries, 1 to 466
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Year           641 non-null    object 
 1   Manufacturer   641 non-null    object 
 2   Model          641 non-null    object 
 3   Vehicle Class  641 non-null    object 
 4   Fuel           641 non-null    object 
 5   Consumption    641 non-null    float64
 6   CO2 (mg/mile)  641 non-null    float64
dtypes: float64(2), object(5)
memory usage: 40.1+ KB


## ETL: `open-meteo-40.74N74.04W51m.csv`

### El archivo que contiene los datos climáticos de la ciudad de New York se extrajo del siguiente [enlace](https://open-meteo.com/en/docs/historical-weather-api)

In [281]:
# Cargamos el archivo de datos
climate = pd.read_csv(r'C:\Users\melan\Desktop\PG 1\ETL -EDA\Data\1. Originales\open-meteo-40.74N74.04W51m.csv')
climate

latitude            longitude  elevation  utc_offset_seconds  \
0            40.738136            -74.04254       51.0            -18000.0   
1                 time  temperature_2m (°C)        NaN                 NaN   
2     2023-01-01T00:00                 10.4        NaN                 NaN   
3     2023-01-01T01:00                 10.3        NaN                 NaN   
4     2023-01-01T02:00                 10.2        NaN                 NaN   
...                ...                  ...        ...                 ...   
8469  2023-12-19T19:00                  NaN        NaN                 NaN   
8470  2023-12-19T20:00                  NaN        NaN                 NaN   
8471  2023-12-19T21:00                  NaN        NaN                 NaN   
8472  2023-12-19T22:00                  NaN        NaN                 NaN   
8473  2023-12-19T23:00                  NaN        NaN                 NaN   

              timezone timezone_abbreviation  
0     America/New_York                   EST  
1                  NaN                   NaN  
2                  NaN                   NaN  
3                  NaN                   NaN  
4                  NaN                   NaN  
...                ...                   ...  
8469               NaN                   NaN  
8470               NaN                   NaN  
8471               NaN                   NaN  
8472               NaN                   NaN  
8473               NaN                   NaN  

[8474 rows x 6 columns]

In [282]:
# Revisamos las columnas
climate.columns

Index(['latitude', 'longitude', 'elevation', 'utc_offset_seconds', 'timezone',
       'timezone_abbreviation'],
      dtype='object')

In [283]:
# Eliminamos las columnas que no nos interesan
eliminar = ['elevation', 'utc_offset_seconds', 'timezone', 'timezone_abbreviation']

climate = climate.drop(eliminar, axis=1)
climate.head(2)

latitude            longitude
0  40.738136            -74.04254
1       time  temperature_2m (°C)

In [284]:
# Eliminamos las dos primeras filas del DataFrame
climate = climate.drop([0,1], axis=0)
climate.reset_index(drop=True)
climate.head(2)

latitude longitude
2  2023-01-01T00:00      10.4
3  2023-01-01T01:00      10.3

In [285]:
climate.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8472 entries, 2 to 8473
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   latitude   8472 non-null   object
 1   longitude  8419 non-null   object
dtypes: object(2)
memory usage: 132.5+ KB


In [288]:
def split_date_time(date_time_str):
    # Suponiendo que la cadena de fecha y hora tiene un formato como 'yyyy-mm-dd HH:MM:SS'
    date_time_parts = date_time_str.split(' ')  # Divide la cadena en partes utilizando un espacio como delimitador
    if len(date_time_parts) == 2:
        date_str, time_str = date_time_parts
        return pd.Series({'Date': date_str, 'Time': time_str})
    else:
        # Manejar el caso de datos incompletos o formato incorrecto
        return pd.Series({'Date': None, 'Time': None})  # Puedes ajustar esto según sea necesario


In [289]:
# Aplicamos la función a la columna "latitude"
climate[['Date', 'Time']] = climate['latitude'].apply(split_date_time).apply(pd.Series)



In [290]:
# Revisamos el resultado
climate.head(2)

latitude longitude  Date  Time
2  2023-01-01T00:00      10.4  None  None
3  2023-01-01T01:00      10.3  None  None

In [291]:
# Eliminamos la columna 'latitude'
climate = climate.drop('latitude', axis=1)
climate.head(2)

longitude  Date  Time
2      10.4  None  None
3      10.3  None  None

In [292]:
# Renombramos la columna 'longitude' a 'Temperature'
climate.rename(columns={'longitude':'Temperature (°C)'}, inplace=True)
climate.head(2)

Temperature (°C)  Date  Time
2             10.4  None  None
3             10.3  None  None

In [293]:
# Eliminamos las filas donde Date es mayor a 2023-09-30
climate = climate[climate['Date'] <= '2023-10-31']
climate.head(2)

Empty DataFrame
Columns: [Temperature (°C), Date, Time]
Index: []

### Guardamos el archivo en formato parquet

In [296]:
climate.to_parquet('C:\\Users\\melan\\Desktop\\PG 1\\ETL -EDA\\Data\\2. Normalizacion\\climate_NY.parquet', index=False)


## DataFrame: `Avg_Fuel Price`

#### Creamos un DataFrame con los datos de los precios de los combustibles en New York, dejando los respectivos enlaces para verificar costes de [Gasoline](https://afdc.energy.gov/fuels/prices.html#:~:text=Gasoline-,%243.72/gallon,-Diesel), [Diesel](https://afdc.energy.gov/fuels/prices.html#:~:text=Diesel-,%244.52/gallon,-Source%3A%20Alternative), [kWh](https://energy5.com/es/encuesta-sobre-los-costos-de-carga-de-vehiculos-electricos-en-diferentes-estados-de-ee-uu-en-jerga-tecnologica#:~:text=Nueva%20York%20%2D%20El%20coste%20medio%20de%20la%20electricidad%20en%20Nueva%20York%20es%20de%200%2C16%20d%C3%B3lares%20por%20kilovatio%2Dhora%2C%20ligeramente%20superior%20al%20promedio%20nacional.), [CNG](https://afdc.energy.gov/fuels/prices.html#:~:text=Natural%20Gas%20(CNG)-,%242.85/GGE,-Liquefied%20Natural%20Gas), [Ethanol e85](https://afdc.energy.gov/fuels/prices.html#:~:text=Ethanol%20(E85)-,%243.05/gallon,-Natural%20Gas%20(CNG)), [Bio-Diesel B20](https://afdc.energy.gov/fuels/prices.html#:~:text=Biodiesel%20(B20)-,%244.42/gallon,-Biodiesel%20(B99%2DB100)) y [Propano LPG](https://afdc.energy.gov/fuels/prices.html#:~:text=Propane-,%243.29/gallon,-Gasoline).

In [297]:
# Creamos un DataFrame que contengan los datos de precios obtenidos de combustibles
fuel_prices = pd.DataFrame({'Fuel': ['Gasoline', 'Diesel', 'Elec', 'CNG', 'E85', 'BD', 'LPG'],
                            'Price (USD)': [3.72, 4.52, 0.16, 2.85, 3.05, 4.42, 3.29],
                            'Measure': ['gallon', 'gallon', 'kWh', 'GGE', 'gallon', 'gallon', 'gallon']})

fuel_prices

Fuel  Price (USD) Measure
0  Gasoline         3.72  gallon
1    Diesel         4.52  gallon
2      Elec         0.16     kWh
3       CNG         2.85     GGE
4       E85         3.05  gallon
5        BD         4.42  gallon
6       LPG         3.29  gallon

### Guardamos el archivo en formato parquet

In [306]:
fuel_prices.to_parquet(r'C:\Users\melan\Desktop\PG 1\ETL -EDA\Data\2. Normalizacion\fuel_prices.parquet', index=False)



## DataFrame: `Edmunds_data`

#### Descargamos datos de vehículos en New York, dejando el respectivo enlace para verificar costes de los mismos en el sitio web [Edmunds](https://www.edmunds.com/)

In [308]:
vehicles_info = pd.read_excel(r'C:\Users\melan\Desktop\PG 1\ETL -EDA\Data\3. Extras\edmunds_data.xlsx')
vehicles_info.head(2)

Year    Make   Model      Fuel Alternative Fuel  City MPG  \
0  2023  Toyota   Camry  Gasoline              NaN        28   
1  2023   Honda  Accord  Gasoline              NaN        27   

   CO2 Emissions (g/mi)  Sound Emission (dB)  Base Price (USD)  Range (mi)  \
0                   229                   74             25075         418   
1                   227                   75             24795         402   

  Vehicle Type  
0        Sedan  
1        Sedan

In [309]:
vehicles_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101 entries, 0 to 100
Data columns (total 11 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   Year                  101 non-null    int64 
 1   Make                  101 non-null    object
 2   Model                 101 non-null    object
 3   Fuel                  101 non-null    object
 4   Alternative Fuel      55 non-null     object
 5   City MPG              101 non-null    int64 
 6   CO2 Emissions (g/mi)  101 non-null    int64 
 7   Sound Emission (dB)   101 non-null    int64 
 8   Base Price (USD)      101 non-null    int64 
 9   Range (mi)            101 non-null    int64 
 10  Vehicle Type          101 non-null    object
dtypes: int64(6), object(5)
memory usage: 8.8+ KB


In [310]:
# Convertimos las columnas 'Make', 'Model', 'Fuel, 'Alternative Fuel' y 'Vehicle Type' a tipo string
vehicles_info['Make'] = vehicles_info['Make'].astype(str)
vehicles_info['Model'] = vehicles_info['Model'].astype(str)
vehicles_info['Fuel'] = vehicles_info['Fuel'].astype(str)
vehicles_info['Alternative Fuel'] = vehicles_info['Alternative Fuel'].astype(str)
vehicles_info['Vehicle Type'] = vehicles_info['Vehicle Type'].astype(str)
vehicles_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101 entries, 0 to 100
Data columns (total 11 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   Year                  101 non-null    int64 
 1   Make                  101 non-null    object
 2   Model                 101 non-null    object
 3   Fuel                  101 non-null    object
 4   Alternative Fuel      101 non-null    object
 5   City MPG              101 non-null    int64 
 6   CO2 Emissions (g/mi)  101 non-null    int64 
 7   Sound Emission (dB)   101 non-null    int64 
 8   Base Price (USD)      101 non-null    int64 
 9   Range (mi)            101 non-null    int64 
 10  Vehicle Type          101 non-null    object
dtypes: int64(6), object(5)
memory usage: 8.8+ KB


In [311]:
# Renombramos columnas
vehicles_info.rename(columns={'Make':'Manufacturer', 'City MPG':'Miles per gallon', 'CO2 Emissions (g/mi)':'CO2 Emission (g/mi)',
                        'Base Price (USD)':'Price (USD)', 'Vehicle Type':'Category'}, inplace=True)

vehicles_info.head(2)

Year Manufacturer   Model      Fuel Alternative Fuel  Miles per gallon  \
0  2023       Toyota   Camry  Gasoline              nan                28   
1  2023        Honda  Accord  Gasoline              nan                27   

   CO2 Emission (g/mi)  Sound Emission (dB)  Price (USD)  Range (mi) Category  
0                  229                   74        25075         418    Sedan  
1                  227                   75        24795         402    Sedan

In [312]:
# Revisamos cantidad de nulos
vehicles_info.isnull().sum()

Year                   0
Manufacturer           0
Model                  0
Fuel                   0
Alternative Fuel       0
Miles per gallon       0
CO2 Emission (g/mi)    0
Sound Emission (dB)    0
Price (USD)            0
Range (mi)             0
Category               0
dtype: int64

In [313]:
vehicles_info['Alternative Fuel'].unique()

array(['nan', 'Electric', 'B20', 'CNG'], dtype=object)

In [314]:
# Reemplazamos los campos con la leyenda 'nan' por 'No'
vehicles_info['Alternative Fuel'] = vehicles_info['Alternative Fuel'].replace(['nan'], 'No')
vehicles_info['Alternative Fuel'].unique()

array(['No', 'Electric', 'B20', 'CNG'], dtype=object)

### Guardamos el archivo en formato parquet

In [323]:
vehicles_info.to_parquet(r'C:\Users\melan\Desktop\PG 1\ETL -EDA\Data\2. Normalizacion\vehicles_info.parquet', index=False)


### **`Observación:` 


Después de completar el proceso de Extracción, Transformación y Carga (ETL) de los conjuntos de datos proporcionados, así como de los datos adicionales descargados, y considerando la estrategia definida para nuestro trabajo, hemos decidido utilizar únicamente los siguientes conjuntos de datos:

air_quality
sound_quality
climate
elec_fuel_charg_stat
fuel_prices
vehicles_info
taxi_zones
Estos conjuntos de datos han sido exportados en formato Parquet a la carpeta 2. Normalizacion. Esta selección se realizó en base a los requisitos y objetivos específicos del proyecto, así como a las necesidades de análisis y modelado de datos.




